lib/Containers/MerkleTree.cpp

```diff
@@ -920,15 +1171,112 @@ std::pair<std::uint64_t, std::uint64_t> MerkleTree<Hasher, BranchingBits, LockSt
+  // not thread-safe, lock buffer from outside
+  if (!_buffer) {
+    return;
+  }
+
+  // validate meta data
+  std::uint64_t maxDepth = meta().maxDepth;
+  std::uint64_t rangeMin = meta().rangeMin;
+  std::uint64_t rangeMax = meta().rangeMax;
+  
+  if (maxDepth < 2) {
+    throw std::invalid_argument("Invalid tree maxDepth");
+  }
+
+  if (rangeMin >= rangeMax) { // || rangeMin == 0) {
+    throw std::invalid_argument("Invalid tree rangeMin / rangeMax");
+  }
+  if (!NumberUtils::isPowerOf2(rangeMax - rangeMin)) {
+    throw std::invalid_argument("Expecting difference between min and max to be power of 2");
+  }
+
+  std::uint64_t previousCount = this->node(0).count;
+  std::uint64_t previousHash = this->node(0).hash;
+
+  for (std::uint64_t d = 1; d <= maxDepth; ++d) {
+    std::uint64_t const offset = nodeCountUpToDepth(d - 1);
+    std::uint64_t count = 0;
+    std::uint64_t hash = 0;
+
+    for (std::uint64_t i = 0; i < nodeCountAtDepth(d); ++i) {
+      Node const& n = this->node(offset + i);
+      count += n.count;
+      hash ^= n.hash;
+
+      TRI_ASSERT(n.count != 0 || n.hash == 0);
+    }
+
+    if (count != previousCount) {
+      throw std::invalid_argument("Inconsistent count values in tree");
+    }
+    previousCount = count;
+    
+    if (hash != previousHash) {
+      throw std::invalid_argument("Inconsistent hash values in tree");
+    }
+  }
+}
+
+template <typename Hasher, std::uint64_t const BranchingBits>
+void MerkleTree<Hasher, BranchingBits>::storeBottomMostCompressed(std::string& output) const {
+  // make a minimum allocation that will be enough for the meta data and a 
+  // few notes. if we need more space, the string can grow as needed
+  TRI_ASSERT(output.empty());
+  output.reserve(64);
+
+  // rangeMin / rangeMax / maxDepth
+  {
+    uint64_t value = basics::hostToLittle(meta().rangeMin);
+    output.append(reinterpret_cast<const char*>(&value), sizeof(uint64_t));
+    
+    value = basics::hostToLittle(meta().rangeMax);
+    output.append(reinterpret_cast<const char*>(&value), sizeof(uint64_t));
+    
+    value = basics::hostToLittle(meta().maxDepth);
+    output.append(reinterpret_cast<const char*>(&value), sizeof(uint64_t));
+  }
+  
+  std::uint64_t maxDepth = meta().maxDepth;
+  std::uint64_t offset = nodeCountUpToDepth(maxDepth - 1);
+  for (std::uint64_t index = 0; index < nodeCountAtDepth(maxDepth); ++index) {
+    Node const& src = this->node(offset + index);
+    TRI_ASSERT(src.count > 0 || src.hash == 0);
+
+    // serialize only the non-zero buckets
+    if (src.count > 0) {
+      // index
+      {
+        uint32_t value = basics::hostToLittle(static_cast<uint32_t>(index));
+        output.append(reinterpret_cast<const char*>(&value), sizeof(uint32_t));
+      }
+      // count / hash
+      {
+        uint64_t value = basics::hostToLittle(src.count);
+        output.append(reinterpret_cast<const char*>(&value), sizeof(uint64_t));
```

Comment:
Since this encoding is only used for trees with a rather small number of entries we could squeeze offset and count into the same 64-bit value. If we keep the cutoff below 16k entries even a 32-bit value would suffice. This would further decrease the required memory.

Reply:
Yes, we can still improve here.
I think it makes sense to improve the bottom-most format in devel when there is time, and then backport it to 3.8 as an additional encoding variant if it is significantly better.